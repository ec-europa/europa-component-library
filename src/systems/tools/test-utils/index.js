const merge = require('deepmerge');
const twing = require('@ecl-twig/ec-storybook/.storybook/environment');

const renderTwigFileAsNode = (file, options) =>
  new Promise((resolve, reject) => {
    try {
      const html = twing.render(file, options);
      const test = document.createElement('test');
      test.innerHTML = html.trim();
      resolve(test);
    } catch (error) {
      reject(error);
    }
  });

module.exports = {
  merge,
  renderTwigFileAsNode,
};
