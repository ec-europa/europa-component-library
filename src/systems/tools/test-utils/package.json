{
  "name": "@ecl-twig/test-utils",
  "version": "2.31.1",
  "description": "A local test helper",
  "main": "index.js",
  "private": true,
  "devDependencies": {
    "deepmerge": "4.2.2"
  }
}
