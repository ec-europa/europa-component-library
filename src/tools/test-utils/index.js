const merge = require('deepmerge');
const twing = require('@ecl/twig-ec-storybook/.storybook/environment');

const renderTwigFileAsNode = (file, options) =>
  new Promise((resolve, reject) => {
    try {
      const html = twing.render(file, options);
      const div = document.createElement('div');
      div.innerHTML = html.trim();
      resolve(div.firstChild);
    } catch (error) {
      reject(error);
    }
  });

module.exports = {
  merge,
  renderTwigFileAsNode,
};
