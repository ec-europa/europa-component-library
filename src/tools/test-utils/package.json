{
  "name": "@ecl/test-utils",
  "version": "3.4.2",
  "description": "A local test helper",
  "main": "index.js",
  "private": true,
  "devDependencies": {
    "deepmerge": "4.2.2"
  }
}
