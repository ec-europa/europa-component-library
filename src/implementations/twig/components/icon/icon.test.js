import { merge, renderTwigFileAsNode } from '@ecl/test-utils';

import brandedIcons from '@ecl/resources-ec-icons/dist/lists/branded.json';
import generalIcons from '@ecl/resources-ec-icons/dist/lists/general.json';
import notificationsIcons from '@ecl/resources-ec-icons/dist/lists/notifications.json';
import uiIcons from '@ecl/resources-ec-icons/dist/lists/ui.json';
import dataBranded from '@ecl/specs-component-icon/demo/data--branded';
import dataNotifications from '@ecl/specs-component-icon/demo/data--notifications';
import dataUi from '@ecl/specs-component-icon/demo/data--ui';

describe('Icon', () => {
  const template = '@ecl/icon/icon.html.twig';
  const render = (params) => renderTwigFileAsNode(template, params);

  describe('Branded', () => {
    brandedIcons.forEach((icon) => {
      test(`- icon ${icon} renders correctly`, () => {
        expect.assertions(1);

        const options = merge(dataBranded, {
          icon: {
            name: icon,
          },
        });

        return expect(render(options)).resolves.toMatchSnapshot();
      });
    });
  });

  describe('Notifications', () => {
    notificationsIcons.forEach((icon) => {
      test(`- icon ${icon} renders correctly`, () => {
        expect.assertions(1);

        const options = merge(dataNotifications, {
          icon: {
            name: icon,
          },
        });

        return expect(render(options)).resolves.toMatchSnapshot();
      });
    });
  });

  describe('General', () => {
    generalIcons.forEach((icon) => {
      test(`- icon ${icon} renders correctly`, () => {
        expect.assertions(1);

        const options = merge(dataUi, {
          icon: {
            name: icon,
            type: 'general',
          },
        });

        return expect(render(options)).resolves.toMatchSnapshot();
      });
    });
  });

  describe('UI', () => {
    uiIcons.forEach((icon) => {
      test(`- icon ${icon} renders correctly`, () => {
        expect.assertions(1);

        const options = merge(dataUi, {
          icon: {
            name: icon,
          },
        });

        return expect(render(options)).resolves.toMatchSnapshot();
      });
    });
  });

  describe('Generic tests - Any icon', () => {
    const options = merge(dataUi, {
      icon: {
        name: generalIcons[0],
        type: 'general',
      },
    });

    test('renders correctly with accessibility content', () => {
      expect.assertions(1);

      const optionsWithAccessibility = merge(options, {
        as_image: true,
        extra_accessibility: {
          title: 'Title',
          title_id: 'example-title',
          description: 'Description',
          description_id: 'example-desc',
        },
      });

      return expect(
        render(optionsWithAccessibility)
      ).resolves.toMatchSnapshot();
    });

    test('renders correctly with extra class names', () => {
      expect.assertions(1);

      const optionsWithExtraClasses = merge(options, {
        extra_classes: 'custom-class custom-class--test',
      });

      return expect(render(optionsWithExtraClasses)).resolves.toMatchSnapshot();
    });

    test('renders correctly with extra attributes', () => {
      expect.assertions(1);

      const optionsWithExtraClasses = merge(options, {
        extra_attributes: [
          { name: 'data-test', value: 'data-test-value' },
          { name: 'data-test-1', value: 'data-test-value-1' },
        ],
      });

      return expect(render(optionsWithExtraClasses)).resolves.toMatchSnapshot();
    });
  });
});
