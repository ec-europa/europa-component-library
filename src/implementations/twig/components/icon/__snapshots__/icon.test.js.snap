// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon Branded - icon facebook renders correctly 1`] = `
<svg
  aria-hidden="true"
  class="ecl-icon ecl-icon--m"
  focusable="false"
>
  <use
    xlink:href="/icons.svg#branded--facebook"
  />
</svg>
`;

exports[`Icon Branded - icon instagram renders correctly 1`] = `
<svg
  aria-hidden="true"
  class="ecl-icon ecl-icon--m"
  focusable="false"
>
  <use
    xlink:href="/icons.svg#branded--instagram"
  />
</svg>
`;

exports[`Icon Branded - icon linkedin renders correctly 1`] = `
<svg
  aria-hidden="true"
  class="ecl-icon ecl-icon--m"
  focusable="false"
>
  <use
    xlink:href="/icons.svg#branded--linkedin"
  />
</svg>
`;

exports[`Icon Branded - icon pinterest renders correctly 1`] = `
<svg
  aria-hidden="true"
  class="ecl-icon ecl-icon--m"
  focusable="false"
>
  <use
    xlink:href="/icons.svg#branded--pinterest"
  />
</svg>
`;

exports[`Icon Branded - icon rss renders correctly 1`] = `
<svg
  aria-hidden="true"
  class="ecl-icon ecl-icon--m"
  focusable="false"
>
  <use
    xlink:href="/icons.svg#branded--rss"
  />
</svg>
`;

exports[`Icon Branded - icon skype renders correctly 1`] = `
<svg
  aria-hidden="true"
  class="ecl-icon ecl-icon--m"
  focusable="false"
>
  <use
    xlink:href="/icons.svg#branded--skype"
  />
</svg>
`;

exports[`Icon Branded - icon twitter renders correctly 1`] = `
<svg
  aria-hidden="true"
  class="ecl-icon ecl-icon--m"
  focusable="false"
>
  <use
    xlink:href="/icons.svg#branded--twitter"
  />
</svg>
`;

exports[`Icon Branded - icon youtube renders correctly 1`] = `
<svg
  aria-hidden="true"
  class="ecl-icon ecl-icon--m"
  focusable="false"
>
  <use
    xlink:href="/icons.svg#branded--youtube"
  />
</svg>
`;

exports[`Icon General - icon audio renders correctly 1`] = `
<svg
  aria-hidden="true"
  class="ecl-icon ecl-icon--m"
  focusable="false"
>
  <use
    xlink:href="/icons.svg#general--audio"
  />
</svg>
`;

exports[`Icon General - icon book renders correctly 1`] = `
<svg
  aria-hidden="true"
  class="ecl-icon ecl-icon--m"
  focusable="false"
>
  <use
    xlink:href="/icons.svg#general--book"
  />
</svg>
`;

exports[`Icon General - icon calendar renders correctly 1`] = `
<svg
  aria-hidden="true"
  class="ecl-icon ecl-icon--m"
  focusable="false"
>
  <use
    xlink:href="/icons.svg#general--calendar"
  />
</svg>
`;

exports[`Icon General - icon copy renders correctly 1`] = `
<svg
  aria-hidden="true"
  class="ecl-icon ecl-icon--m"
  focusable="false"
>
  <use
    xlink:href="/icons.svg#general--copy"
  />
</svg>
`;

exports[`Icon General - icon digital renders correctly 1`] = `
<svg
  aria-hidden="true"
  class="ecl-icon ecl-icon--m"
  focusable="false"
>
  <use
    xlink:href="/icons.svg#general--digital"
  />
</svg>
`;

exports[`Icon General - icon edit renders correctly 1`] = `
<svg
  aria-hidden="true"
  class="ecl-icon ecl-icon--m"
  focusable="false"
>
  <use
    xlink:href="/icons.svg#general--edit"
  />
</svg>
`;

exports[`Icon General - icon faq renders correctly 1`] = `
<svg
  aria-hidden="true"
  class="ecl-icon ecl-icon--m"
  focusable="false"
>
  <use
    xlink:href="/icons.svg#general--faq"
  />
</svg>
`;

exports[`Icon General - icon feedback renders correctly 1`] = `
<svg
  aria-hidden="true"
  class="ecl-icon ecl-icon--m"
  focusable="false"
>
  <use
    xlink:href="/icons.svg#general--feedback"
  />
</svg>
`;

exports[`Icon General - icon file renders correctly 1`] = `
<svg
  aria-hidden="true"
  class="ecl-icon ecl-icon--m"
  focusable="false"
>
  <use
    xlink:href="/icons.svg#general--file"
  />
</svg>
`;

exports[`Icon General - icon gear renders correctly 1`] = `
<svg
  aria-hidden="true"
  class="ecl-icon ecl-icon--m"
  focusable="false"
>
  <use
    xlink:href="/icons.svg#general--gear"
  />
</svg>
`;

exports[`Icon General - icon generic-lang renders correctly 1`] = `
<svg
  aria-hidden="true"
  class="ecl-icon ecl-icon--m"
  focusable="false"
>
  <use
    xlink:href="/icons.svg#general--generic-lang"
  />
</svg>
`;

exports[`Icon General - icon global renders correctly 1`] = `
<svg
  aria-hidden="true"
  class="ecl-icon ecl-icon--m"
  focusable="false"
>
  <use
    xlink:href="/icons.svg#general--global"
  />
</svg>
`;

exports[`Icon General - icon growth renders correctly 1`] = `
<svg
  aria-hidden="true"
  class="ecl-icon ecl-icon--m"
  focusable="false"
>
  <use
    xlink:href="/icons.svg#general--growth"
  />
</svg>
`;

exports[`Icon General - icon hamburger renders correctly 1`] = `
<svg
  aria-hidden="true"
  class="ecl-icon ecl-icon--m"
  focusable="false"
>
  <use
    xlink:href="/icons.svg#general--hamburger"
  />
</svg>
`;

exports[`Icon General - icon image renders correctly 1`] = `
<svg
  aria-hidden="true"
  class="ecl-icon ecl-icon--m"
  focusable="false"
>
  <use
    xlink:href="/icons.svg#general--image"
  />
</svg>
`;

exports[`Icon General - icon infographic renders correctly 1`] = `
<svg
  aria-hidden="true"
  class="ecl-icon ecl-icon--m"
  focusable="false"
>
  <use
    xlink:href="/icons.svg#general--infographic"
  />
</svg>
`;

exports[`Icon General - icon language renders correctly 1`] = `
<svg
  aria-hidden="true"
  class="ecl-icon ecl-icon--m"
  focusable="false"
>
  <use
    xlink:href="/icons.svg#general--language"
  />
</svg>
`;

exports[`Icon General - icon livestreaming renders correctly 1`] = `
<svg
  aria-hidden="true"
  class="ecl-icon ecl-icon--m"
  focusable="false"
>
  <use
    xlink:href="/icons.svg#general--livestreaming"
  />
</svg>
`;

exports[`Icon General - icon location renders correctly 1`] = `
<svg
  aria-hidden="true"
  class="ecl-icon ecl-icon--m"
  focusable="false"
>
  <use
    xlink:href="/icons.svg#general--location"
  />
</svg>
`;

exports[`Icon General - icon log-in renders correctly 1`] = `
<svg
  aria-hidden="true"
  class="ecl-icon ecl-icon--m"
  focusable="false"
>
  <use
    xlink:href="/icons.svg#general--log-in"
  />
</svg>
`;

exports[`Icon General - icon logged-in renders correctly 1`] = `
<svg
  aria-hidden="true"
  class="ecl-icon ecl-icon--m"
  focusable="false"
>
  <use
    xlink:href="/icons.svg#general--logged-in"
  />
</svg>
`;

exports[`Icon General - icon multiple-files renders correctly 1`] = `
<svg
  aria-hidden="true"
  class="ecl-icon ecl-icon--m"
  focusable="false"
>
  <use
    xlink:href="/icons.svg#general--multiple-files"
  />
</svg>
`;

exports[`Icon General - icon organigram renders correctly 1`] = `
<svg
  aria-hidden="true"
  class="ecl-icon ecl-icon--m"
  focusable="false"
>
  <use
    xlink:href="/icons.svg#general--organigram"
  />
</svg>
`;

exports[`Icon General - icon print renders correctly 1`] = `
<svg
  aria-hidden="true"
  class="ecl-icon ecl-icon--m"
  focusable="false"
>
  <use
    xlink:href="/icons.svg#general--print"
  />
</svg>
`;

exports[`Icon General - icon regulation renders correctly 1`] = `
<svg
  aria-hidden="true"
  class="ecl-icon ecl-icon--m"
  focusable="false"
>
  <use
    xlink:href="/icons.svg#general--regulation"
  />
</svg>
`;

exports[`Icon General - icon search renders correctly 1`] = `
<svg
  aria-hidden="true"
  class="ecl-icon ecl-icon--m"
  focusable="false"
>
  <use
    xlink:href="/icons.svg#general--search"
  />
</svg>
`;

exports[`Icon General - icon share renders correctly 1`] = `
<svg
  aria-hidden="true"
  class="ecl-icon ecl-icon--m"
  focusable="false"
>
  <use
    xlink:href="/icons.svg#general--share"
  />
</svg>
`;

exports[`Icon General - icon spinner renders correctly 1`] = `
<svg
  aria-hidden="true"
  class="ecl-icon ecl-icon--m"
  focusable="false"
>
  <use
    xlink:href="/icons.svg#general--spinner"
  />
</svg>
`;

exports[`Icon General - icon spreadsheet renders correctly 1`] = `
<svg
  aria-hidden="true"
  class="ecl-icon ecl-icon--m"
  focusable="false"
>
  <use
    xlink:href="/icons.svg#general--spreadsheet"
  />
</svg>
`;

exports[`Icon General - icon tag renders correctly 1`] = `
<svg
  aria-hidden="true"
  class="ecl-icon ecl-icon--m"
  focusable="false"
>
  <use
    xlink:href="/icons.svg#general--tag"
  />
</svg>
`;

exports[`Icon General - icon video renders correctly 1`] = `
<svg
  aria-hidden="true"
  class="ecl-icon ecl-icon--m"
  focusable="false"
>
  <use
    xlink:href="/icons.svg#general--video"
  />
</svg>
`;

exports[`Icon Generic tests - Any icon renders correctly with accessibility content 1`] = `
<svg
  aria-hidden="false"
  aria-labelledby="example-title example-desc"
  class="ecl-icon ecl-icon--m"
  focusable="false"
  role="img"
>
  <title
    id="example-title"
  >
    Title
  </title>
  <desc
    id="example-desc"
  >
    Description
  </desc>
  <use
    xlink:href="/icons.svg#general--audio"
  />
</svg>
`;

exports[`Icon Generic tests - Any icon renders correctly with extra attributes 1`] = `
<svg
  aria-hidden="true"
  class="ecl-icon ecl-icon--m"
  data-test="data-test-value"
  data-test-1="data-test-value-1"
  focusable="false"
>
  <use
    xlink:href="/icons.svg#general--audio"
  />
</svg>
`;

exports[`Icon Generic tests - Any icon renders correctly with extra class names 1`] = `
<svg
  aria-hidden="true"
  class="ecl-icon ecl-icon--m custom-class custom-class--test"
  focusable="false"
>
  <use
    xlink:href="/icons.svg#general--audio"
  />
</svg>
`;

exports[`Icon Notifications - icon error renders correctly 1`] = `
<svg
  aria-hidden="true"
  class="ecl-icon ecl-icon--m"
  focusable="false"
>
  <use
    xlink:href="/icons.svg#notifications--error"
  />
</svg>
`;

exports[`Icon Notifications - icon information renders correctly 1`] = `
<svg
  aria-hidden="true"
  class="ecl-icon ecl-icon--m"
  focusable="false"
>
  <use
    xlink:href="/icons.svg#notifications--information"
  />
</svg>
`;

exports[`Icon Notifications - icon success renders correctly 1`] = `
<svg
  aria-hidden="true"
  class="ecl-icon ecl-icon--m"
  focusable="false"
>
  <use
    xlink:href="/icons.svg#notifications--success"
  />
</svg>
`;

exports[`Icon Notifications - icon warning renders correctly 1`] = `
<svg
  aria-hidden="true"
  class="ecl-icon ecl-icon--m"
  focusable="false"
>
  <use
    xlink:href="/icons.svg#notifications--warning"
  />
</svg>
`;

exports[`Icon UI - icon check renders correctly 1`] = `
<svg
  aria-hidden="true"
  class="ecl-icon ecl-icon--m"
  focusable="false"
>
  <use
    xlink:href="/icons.svg#ui--check"
  />
</svg>
`;

exports[`Icon UI - icon check-filled renders correctly 1`] = `
<svg
  aria-hidden="true"
  class="ecl-icon ecl-icon--m"
  focusable="false"
>
  <use
    xlink:href="/icons.svg#ui--check-filled"
  />
</svg>
`;

exports[`Icon UI - icon close renders correctly 1`] = `
<svg
  aria-hidden="true"
  class="ecl-icon ecl-icon--m"
  focusable="false"
>
  <use
    xlink:href="/icons.svg#ui--close"
  />
</svg>
`;

exports[`Icon UI - icon close-filled renders correctly 1`] = `
<svg
  aria-hidden="true"
  class="ecl-icon ecl-icon--m"
  focusable="false"
>
  <use
    xlink:href="/icons.svg#ui--close-filled"
  />
</svg>
`;

exports[`Icon UI - icon corner-arrow renders correctly 1`] = `
<svg
  aria-hidden="true"
  class="ecl-icon ecl-icon--m"
  focusable="false"
>
  <use
    xlink:href="/icons.svg#ui--corner-arrow"
  />
</svg>
`;

exports[`Icon UI - icon download renders correctly 1`] = `
<svg
  aria-hidden="true"
  class="ecl-icon ecl-icon--m"
  focusable="false"
>
  <use
    xlink:href="/icons.svg#ui--download"
  />
</svg>
`;

exports[`Icon UI - icon external renders correctly 1`] = `
<svg
  aria-hidden="true"
  class="ecl-icon ecl-icon--m"
  focusable="false"
>
  <use
    xlink:href="/icons.svg#ui--external"
  />
</svg>
`;

exports[`Icon UI - icon fullscreen renders correctly 1`] = `
<svg
  aria-hidden="true"
  class="ecl-icon ecl-icon--m"
  focusable="false"
>
  <use
    xlink:href="/icons.svg#ui--fullscreen"
  />
</svg>
`;

exports[`Icon UI - icon minus renders correctly 1`] = `
<svg
  aria-hidden="true"
  class="ecl-icon ecl-icon--m"
  focusable="false"
>
  <use
    xlink:href="/icons.svg#ui--minus"
  />
</svg>
`;

exports[`Icon UI - icon plus renders correctly 1`] = `
<svg
  aria-hidden="true"
  class="ecl-icon ecl-icon--m"
  focusable="false"
>
  <use
    xlink:href="/icons.svg#ui--plus"
  />
</svg>
`;

exports[`Icon UI - icon rounded-arrow renders correctly 1`] = `
<svg
  aria-hidden="true"
  class="ecl-icon ecl-icon--m"
  focusable="false"
>
  <use
    xlink:href="/icons.svg#ui--rounded-arrow"
  />
</svg>
`;

exports[`Icon UI - icon solid-arrow renders correctly 1`] = `
<svg
  aria-hidden="true"
  class="ecl-icon ecl-icon--m"
  focusable="false"
>
  <use
    xlink:href="/icons.svg#ui--solid-arrow"
  />
</svg>
`;
